packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -277,6 +293,9 @@ export function createVM<HostNode, HostElement>(
         oar: create(null),
         cmpTemplate: null,
 
+        renderMode,
```

Comment:
Can we remove this and use `def.renderMode` or `def.shadowMode` everywhere else?

Reply:
My mental model is that what goes in the def is constant and is never a function of the environment. That would mean that the render mode can be stored in the def but the shadow mode couldn't, but it feels weird to store these values in different places.

Reply:
I like that mental model, in def we keep the desired value (sort of speak), while in the vm we have the actual value depending on the environment conditions, parent state, etc.

Reply:
I also like this mental model. 

That said the shadow mode is in the gray zone. For synthetic shadow to work, the polyfill has to be evaluated prior to the engine and any LWC component code. Also once the polyfill evaluated it isn't possible to unpatch the API.  Because of this, it should be safe to assume that `isNativeShadowRootDefined` is constant for the lifetime of the engine.

All of this to say that I am fine with either approach.

Reply:
Makes sense. I'll move it to the def like @abdulsattar was implying.

Reply:
Merging this as-is since there is talk of making the shadow mode a function of the component hierarchy. We can make adjustments later as needed.